### PR TITLE
Add fhir-client and EHR login capability

### PR DESCRIPTION
**Pull request by jadh4v**
This PR does the following:
- Add fhirClient-js as a dependency
- Add a new vue Component (EHRDataBrowser.vue) which adds a new tab in Volview for fetching and viewing electronic patient records (EHR).
- Add html pages for cerner standalone app launcher and redirect page (https://code-console.cerner.com/).
- Make Volview a multi-page app so that the EHR app launcher and redirect html pages can be hosted within the same vite server rather than hosting a separate http server for the login process.

### CODE REVIEW

File: vite.config.ts [hunk 1]
```diff
@@ -37,6 +37,11 @@ export default defineConfig({
   build: {
     outDir: distDir,
     rollupOptions: {
+      input: {
+        main: resolve(__dirname, './index.html'),
+        ehrLaunch: resolve(__dirname, './lungair/cerner-app-launch/launch.html'),
+        ehrRedirect: resolve(__dirname, './lungair/cerner-app-launch/index.html'),
+      },
```
Comment (jadh4v) [inline vite.config.ts:44]: @ebrahimebrahim This allows us to use Volview's vite server for our cerner app launch as discussed earlier. So now we don't need that additional http server 😃.

**Comment by ebrahimebrahim** [comment]
While following the instructions in the README for setting up the vite proxy server, I ran into issues at the `npm install` command. Here is the relevant snippet from the npm log:

```txt
74 verbose Linux 5.15.0-78-generic
75 verbose node v18.14.0
76 verbose npm  v9.3.1
77 error code 1
78 error path /home/ebrahim/lungair-web-application/lungair/orthanc-proxy/node_modules/esbuild
79 error command failed
80 error command sh -c node install.js
81 error /home/ebrahim/lungair-web-application/lungair/orthanc-proxy/node_modules/esbuild/install.js:132
81 error     throw new Error(`Expected ${JSON.stringify(versionFromPackageJSON)} but got ${JSON.stringify(stdout)}`);
81 error           ^
81 error
81 error Error: Expected "0.18.17" but got "0.17.19"
81 error     at validateBinaryVersion (/home/ebrahim/lungair-web-application/lungair/orthanc-proxy/node_modules/esbuild/install.js:132:11)
81 error     at /home/ebrahim/lungair-web-application/lungair/orthanc-proxy/node_modules/esbuild/install.js:285:5
81 error
81 error Node.js v18.14.0
82 verbose exit 1
```

**Comment by ebrahimebrahim** [comment]
Using the _LungAIR EHR_ tab I got the Cerner popup and logged in as one of the test patients, then got the patient dump in the frame. **Really neat!**

When I do it again without refreshing, hitting the login button and logging in as a different patient, the displayed data did not change. Ultimately we are not going for a patient-facing app, so I am not sure if we should worry about this... but probably we should get it right anyway to ensure our early designs can support what we eventually want to do.

**Comment by ebrahimebrahim** [comment]
> Can you try deleting the existing `node_modules` directory, delete `package-lock.json` and try `npm install` again.

That worked (for the `lungair/orthanc-proxy` node project -- deleting `package-lock.json` for the main (volview) node project causes it to fail to build on my machine)

File: lungair/README.md [hunk 2]
```diff
@@ -1,13 +1,78 @@
...
 
+## Instructions to build and run the test server
+Following are instructions to build and run an instance of LungAir web application with
+locally hosted DICOMWeb server, electronics health records (EHR) server, and VolView's
+backend Python server for a connected deep learning pipeline.
+
+### Prerequisites
+- _Node-js_ version >= `18.14.0` (might work with lower versions, but not tested)
+- _Python_ >= `3.9.12` (might work with lower versions, but not tested)
+- Download and install [Orthanc Server](https://www.orthanc-server.com/download.php)
```
Comment (ebrahimebrahim) [inline lungair/README.md:16]: Configuring Orthanc to work with this on linux was nontrivial.
So I might soon add a section explaining how I got this working with docker. Different PR.

File: lungair/README.md [hunk 3]
```diff
@@ -1,13 +1,78 @@
...
+
+   Go to the _LungAir EHR_ tab in VolView. The login process is currently hard-coded for
+   performing a Cerner Code app launch. Click the login and enter credentials in the
+   pop-up window. See this [test data document](https://docs.google.com/document/d/10RnVyF1etl_17pyCyK96tyhUWRbrTyEcqpwzW-Z-Ybs/edit)
+   for credentials to use on Oracle's public test server.
+
+6. VolView Python backend server.
+
+   VolView's python server can be used to run backend jobs such as deep learning inference pipelines.
+   Start an instance of the python server by following the [quick-start guide](../documentation/content/doc/server-dev.md#starting-the-server).
```
Comment (ebrahimebrahim) [inline lungair/README.md:69]: I followed the quick start guide and started an instance of the python server.
That ran without errors but I didn't see any evidence of the volview viewer talking to the python server.
The linked guide says

> Navigate to the "Remote Functions" tab

but there is no such tab for me.

**Review by ebrahimebrahim — COMMENTED** [review]
- commit the `.env` file that is needed for volview to find
- update the package-lock for the vite proxy server

File: lungair/README.md (hunk 2 above)
Comment (jadh4v) [reply lungair/README.md:16]: Just tested that using node v18.17.x works well for both Windows/Ubuntu. I have updated the package-lock.json and the README file accordingly.

**Comment by jadh4v** [comment]
> Using the _LungAIR EHR_ tab I got the Cerner popup and logged in as one of the test patients, then got the patient dump in the frame. **Really neat!**
> 
> When I do it again without refreshing, hitting the login button and logging in as a different patient, the displayed data did not change. Ultimately we are not going for a patient-facing app, so I am not sure if we should worry about this... but probably we should get it right anyway to ensure our early designs can support what we eventually want to do.

It looks like this is a weird use-case for `fhirclient` because it doesn't work with multiple authorization calls without refreshing the entire page first. It is fixable by patching the `fhirclient`, but I would wait to see if this is really a required flow. When working on provider facing app, we may not need to login multiple times, or just trigger a full refresh for another login.

**Comment by ebrahimebrahim** [comment]
I ran a fresh test with the newly updated `package-lock` and `.env` and it worked quite well. Uncommenting the server module in `ModulePanel.vue` also worked and I was able to talk to the python server from the application :tada:

**Comment by jadh4v** [comment]
> > It looks like this is a weird use-case for `fhirclient` because it doesn't work with multiple authorization calls without refreshing the entire page first. It is fixable by patching the `fhirclient`, but I would wait to see if this is really a required flow. When working on provider facing app, we may not need to login multiple times, or just trigger a full refresh for another login.
> 
> I see, agreed. It's not a required flow at all. Just wanted to be sure we weren't doing something wrong, and it seems we are not.

To document the findings:
Since v2.x, [fhirclient](https://github.com/smart-on-fhir/client-js) doesn't support passing of `code` and `state` values as function parameters when calling `FHIR.oauth2.Ready()`. This means that these values are read by the FHIR module through the global `location` variable ([see here](https://github.com/smart-on-fhir/client-js/blob/8f64b770dbcd0abd30646e239cd446dfa4d831f6/src/adapters/BrowserAdapter.ts#L93-L99)).
Currently, the global variable is only fetched the first time and the url string is cached into a member variable of BrowserAdapter. For future calls, instead of re-reading from the window's actual `location`, the cached value is used. This prevents the second login from succeeding. And, currently there is no public function exposed that could be used to reset the state of FHIR module. Refreshing the page is the only way to do this.

I am not certain if this is a bug in `fhir-client` or if this is simply not considered a valid use case where an app could allow a second login without resetting state.

**Pull request merged**